src/collector/top_score_collector.rs

```diff
@@ -736,9 +722,82 @@ impl SegmentCollector for TopScoreSegmentCollector {
     }
 }
 
+/// Fast TopN Computation
+///
+/// For TopN == 0, it will be relative expensive.
+pub struct TopNComputer<Score, DocId> {
+    buffer: Vec<ComparableDoc<Score, DocId>>,
+    top_n: usize,
+    pub(crate) threshold: Option<Score>,
+}
+
+impl<Score, DocId> TopNComputer<Score, DocId>
+where
+    Score: PartialOrd + Clone,
+    DocId: Ord + Clone,
+{
+    /// Create a new `TopNComputer`.
+    /// Internally it will allocate a buffer of size `2 * top_n`.
+    pub fn new(top_n: usize) -> Self {
+        let vec_cap = top_n.max(1) * 2;
+        TopNComputer {
+            buffer: Vec::with_capacity(vec_cap),
+            top_n,
+            threshold: None,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn push(&mut self, doc: ComparableDoc<Score, DocId>) {
+        if let Some(last_median) = self.threshold.clone() {
+            if doc.feature < last_median {
+                return;
+            }
+        }
+        if self.buffer.len() == self.buffer.capacity() {
+            let median = self.truncate_median();
+            self.threshold = Some(median);
+        }
+
+        // This is faster since it avoids the buffer resizing to be inlined from vec.push()
+        // (this is in the hot path)
+        // TODO: Replace with `push_within_capacity` when it's stabilized
+        let uninit = self.buffer.spare_capacity_mut();
+        // This cannot panic, because we truncate_median will at least remove one element, since
+        // the min capacity is 2.
+        uninit[0].write(doc);
+        // This is safe because it would panic in the line above
+        unsafe {
+            self.buffer.set_len(self.buffer.len() + 1);
+        }
+    }
+
+    #[inline(never)]
+    fn truncate_median(&mut self) -> Score {
+        let len = self.buffer.len();
+        // Use select_nth_unstable to find the median score
+        let (_, median_el, _) = self.buffer.select_nth_unstable(len / 2);
```

Comment:
It is more solid to write it use `self.top_n` here I think.
```suggestion
        let (_, median_el, _) = self.buffer.select_nth_unstable(self.top_n);
```